### PR TITLE
docs(trade): TradeWidget architecture docs

### PR DESCRIPTION
**Pull request by shoom3301**
# Summary

Migrated TradeWidget architecture docs from https://www.notion.so/cownation/Trade-widgets-architecture-69c447658f8a48189afeabcbb9589cf1

### CODE REVIEW

**Review by alfetopito — APPROVED** [review]
👏

File: src/cow-react/modules/trade/docs/README.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# Trade module
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/cow-react/modules/trade/docs/README.md:1]: Also, if this is the module, in the example I followed the pattern AAAA-module for both the README and the diagram
Reason is to search easily by open files by name.

image.png

File: src/cow-react/modules/trade/docs/README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+   5. Token approving
+2. Simultaneously, `TradeWidget` allows customizing a widget layout using `TradeWidgetSlots` ([ref](../../../modules/trade/containers/TradeWidget/index.tsx#L33)) and other parameters.
+3. `useTradeState()` hook serves to **generalize** trade states. It returns the current trade widgets raw state depending on what widget is open now. *We take as a fact that we can have only one trade widget on the page at one point in time.*
+   1. This hook is supposed to be a base for **shared trade services** (by service I mean component, hook, updater, atom, etc.)
+4. `Shared trade services` - these are tools that allow you to work with any trade widgets without requiring you to know the type of this widget
+   1. Example: `useIsWrapOrUnwrap()` - the hook returns true when a trade state is ETH/WETH or WETH/ETH. And you can call it from everywhere in the App and it will give the result relative to the current widget (swap/limit orders/advanced orders)
+
+## TradeWidget structure diagram (sub-modules)
+
+![Structure](./structure.drawio.svg)
```
Comment (anxolin) [inline src/cow-react/modules/trade/docs/README.md:30]: It would be good to set the background, in darkmode i can't see the arrows porperly image.png

File: src/cow-react/modules/trade/docs/README.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+Since we have at least three trade widgets we need to unify their logic and avoid code duplication  ([DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)). To do this, we need to define what is specific and what is common.
+At the same time, strike a balance between code cohesion and solution scalability.
+
+## Shared language
+
+- `TradeWidget` - Widget to create CoW orders. Its configuration can create many different ways to create orders, for example Swap, Limit Orders and Advanced orders will be instances of it
+- `TradeRawState` - DTO derived from URL (`/1/USDT/WETH?recipient=0x000`) and persisted into localStorage ([ref](../../../modules/trade/types/TradeRawState.ts))
+- `TradeDerivedState` - a model that contains `CurrencyAmount` and `Currency` instances. This state is derived from TradeRawState ([ref](../../../modules/trade/types/TradeDerivedState.ts))
+
```
Comment (anxolin) [inline src/cow-react/modules/trade/docs/README.md:13]: Some small suggestions

```suggestion
- `Trade widget`: Widget to create CoW orders. Some examples would be the `SwapWidget` or `LimitOrdersWidget`. Normally widget implementations will make use of `TradeWidget`. 
- `AbstractTradeWidget`: Abstraction of a Trade Widget, it contains most of the shared parts of the widgets: main layout, sell token, the recipient and some modals. It will have in its parametrization some slots where other trade widgets will be able to inject components i.e. `LimitOrdersWidget` will be able to add the limit price.
- `AbstractTradeUrlState`: DTO derived from URL (`/1/USDT/WETH?recipient=0x000`) and persisted into localStorage ([ref](../../../modules/trade/types/TradeRawState.ts))
- `AbstractTradeState`: A model that contains `CurrencyAmount` and `Currency` instances. This state is derived from TradeRawState ([ref](../../../modules/trade/types/TradeDerivedState.ts))
```

File: src/cow-react/modules/trade/docs/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+## Concept of TradeWidget
+
+1. Each trade widget: [SwapWidget](../../../modules/swap/containers/SwapWidget/index.tsx), [LimitOrdersWidget](../../../modules/limitOrders/containers/LimitOrdersWidget/index.tsx), [AdvancedOrdersWidget](../../../modules/advancedOrders/containers/AdvancedOrdersWidget/index.tsx) uses [TradeWidget](../../../modules/trade/containers/TradeWidget/index.tsx) as a skeleton to get out of the box:
```
Comment (anxolin) [inline src/cow-react/modules/trade/docs/README.md:16]: For the chart i have some comments, but I think is good enough to leave it like this for now. 

I specially like the trade widget extenssion

File: src/cow-react/modules/trade/docs/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+3. `useTradeState()` hook serves to **generalize** trade states. It returns the current trade widgets raw state depending on what widget is open now. *We take as a fact that we can have only one trade widget on the page at one point in time.*
+   1. This hook is supposed to be a base for **shared trade services** (by service I mean component, hook, updater, atom, etc.)
+4. `Shared trade services` - these are tools that allow you to work with any trade widgets without requiring you to know the type of this widget
+   1. Example: `useIsWrapOrUnwrap()` - the hook returns true when a trade state is ETH/WETH or WETH/ETH. And you can call it from everywhere in the App and it will give the result relative to the current widget (swap/limit orders/advanced orders)
+
+## TradeWidget structure diagram (sub-modules)
+
+![Structure](./structure.drawio.svg)
+
+## State diagram
```
Comment (anxolin) [inline src/cow-react/modules/trade/docs/README.md:32]: "Abstraction Diagram" I think is pointing out an issue with our model

This is an abstraction that knows all the implementations.

This module in order to be clean, shouldn't have logic related to limit orders, etc.

**Review by anxolin — COMMENTED** [review]
This is great!

I'm fine merging this, if the objective is to document what we have. 

Although if the objective is to get an agreement on terminology and the architecture/design of this, i still find some issues with our architecture. 

The most important feedback I can give is, I would like that if `trade` module is just an abstraction facilitating some shared logic among many other widgets, I think it should know nothing about its depedant modules. 

In other words,  `limitOrders` will depend on `trade` module, but not the other way around.

File: src/cow-react/modules/trade/docs/README.md (hunk 5 above)
Comment (shoom3301) [reply src/cow-react/modules/trade/docs/README.md]: 100% agree, it's a breach of the Dependency inversion principle.
I have an idea of how to solve it:
- create an intermediate atom `tradeStateAtom`
- in each trade widget fill in the `tradeStateAtom`
- use `tradeStateAtom` instead of `useTradeState()` for common services

image

File: src/cow-react/modules/trade/docs/README.md (hunk 2 above)
Comment (shoom3301) [reply src/cow-react/modules/trade/docs/README.md]: Fixed

File: src/cow-react/modules/trade/docs/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shoom3301) [reply src/cow-react/modules/trade/docs/README.md]: Thank you! Fixed

File: src/cow-react/modules/trade/docs/README.md (hunk 3 above)
Comment (shoom3301) [reply src/cow-react/modules/trade/docs/README.md]: Great suggestions! Thank you!
It doesn't fully match with the current implementation, if you don't mind I will keep it as it is and will update the doc after code changes.

Comment (shoom3301) [reply src/cow-react/modules/trade/docs/README.md]: Created an issue: https://github.com/cowprotocol/cowswap/issues/2556

**Comment by shoom3301** [comment]
@anxolin Thanks a lot for the really deep review!
I 100% agree with your comments, created [an issue](https://github.com/cowprotocol/cowswap/issues/2556) to improve it.

File: src/cow-react/modules/trade/docs/README.md (hunk 5 above)
Comment (anxolin) [reply src/cow-react/modules/trade/docs/README.md]: that looks better. Still, I would avoid depending on atoms from other modules. I would abstract away that if possible. 
so basically the arrows should be IMO to the hook, and not to the state directly.
